Zend/zend_enum.c

```diff
@@ -251,11 +241,43 @@ static void zend_enum_from_base(INTERNAL_FUNCTION_PARAMETERS, bool try)
 	zval *case_zv = &c->value;
 	if (Z_TYPE_P(case_zv) == IS_CONSTANT_AST) {
 		if (zval_update_constant_ex(case_zv, c->ce) == FAILURE) {
-			RETURN_THROWS();
+			return FAILURE;
 		}
 	}
 
-	ZVAL_COPY(return_value, case_zv);
+	*result = Z_OBJ_P(case_zv);
+	return SUCCESS;
+}
+
+static void zend_enum_from_base(INTERNAL_FUNCTION_PARAMETERS, bool try)
+{
+	zend_class_entry *ce = execute_data->func->common.scope;
+	zend_string *string_key = NULL;
+	zend_long long_key = 0;
+
+	if (ce->enum_backing_type == IS_LONG) {
+		ZEND_PARSE_PARAMETERS_START(1, 1)
+			Z_PARAM_LONG(long_key)
+		ZEND_PARSE_PARAMETERS_END();
+	} else {
+		ZEND_PARSE_PARAMETERS_START(1, 1)
+			Z_PARAM_STR(string_key)
+		ZEND_PARSE_PARAMETERS_END();
+
+		ZEND_ASSERT(ce->enum_backing_type == IS_STRING);
```

Comment:
Move this up

Reply:
I tracked it down to `zend_jit_needs_arg_dtor`. The problem seems to be that `zend_jit_needs_arg_dtor` skips the dtor of parameters if (under normal circumstances) no coercion happens (`ZEND_TYPE_FULL_MASK(arg_info->type) & (1u << type))`). `from`/`tryFrom` are special in that the accepted types depend on the enum type. For string enums, long will get coerced to a string, even though the arg type doesn't indicate that. Anyway, this is something that needs to be fixed in those methods.

Reply:
Will be fixed in https://github.com/php/php-src/pull/8633.